### PR TITLE
Adds back-end support for Business Membership

### PR DESCRIPTION
**Pull request by x110dc**
#### What's this PR do?

- Adds back-end support for Business Membership
- Adds lots of doc strings and comments
- Stops hard-coding record type IDs, time zones and the SF API version
- Adds metadata to the Stripe payment record
- Lazily invokes the SF API
- Adds ability to mass update SF records

#### Why are we doing this? How does it help us?

- We need support for Business Membership
- lazily calling the Salesforce API means fewer calls against our quota
- fewer hard-coded things means it's easier for others to use without changing code

#### How should this be manually tested?

Since this is just the backend portion you'll need to use curl or similar. You'll need to get a test customer ID:

`curl -d "business_name=Donkey&website=https://chicken.com&zipcode=&amount=35&customerId=<customer ID>&description=The-Texas&first_name=A&installment_period=monthly&installments=None&last_name=B&openended_status=Open&pay_fees_value=True&stripeEmail=foo@baz.com&reason=I-love-the-trib" -H "Content-Type: application/x-www-form-urlencoded" -X POST http://local.texastribune.org/bmcharge`

#### Have automated tests been added?

No, but the existing ones still pass! RDOs and Opportunities are still basically the same, and the rest is interaction with Salesforce.

#### Has this been tested on mobile?

N/A

#### Are there performance implications?

There are no frontend changes, but things on the backend may be faster because there should be fewer API calls.

#### What are the relevant tickets?

https://3.basecamp.com/3098728/buckets/736178/todos/1264755101

and

https://3.basecamp.com/3098728/buckets/9070027/todos/1302686356

#### How should this change be communicated to end users?

Unnecessary until the frontend parts are in place.

#### Next steps?

Test with the front end code in place.

#### Smells?

I'm not sure how to handle the results from an SF composite API call. It's supposed to have an ID so you know which record failed but for every test I did it was missing. Also not sure whether to make that all-or-none or leave that configurable per-call.

There's some non-DRY code in npsp.py that I need to factor out next.

Setting the record type ID works for non-open-ended RDOs but for open-ended it'll need to be handled on the Salesforce side.

The code for `construct_slack_message` is ugly because of the different variations.

#### Has the relevant documentation/wiki been updated?

Yes, the wiki has been updated.

#### Technical debt note

More docs/comments so less debt.

### CODE REVIEW

**Comment by AndrewGibson27** [comment]
I'm guessing this an issue with my environment, but I got the following trace when doing a normal transaction from `/donate`. This was after a fresh container build, then a `make restart`:
```
[2018-10-04 16:04:11,950: INFO/ForkPoolWorker-2] response.text:[ {
  "errorCode" : "NOT_FOUND",
  "message" : "The requested resource does not exist"
} ]
[2018-10-04 16:04:11,955: ERROR/ForkPoolWorker-2] Task app.add_donation[fb874b70-41ba-4071-b67d-4105a1a883fb] raised unexpected: SalesforceException('Expected 200 but got 404',)
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/celery/app/trace.py", line 382, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/celery/app/trace.py", line 641, in __protected_call__
    return self.run(*args, **kwargs)
  File "/app/app.py", line 411, in add_donation
    email=email, first_name=first_name, last_name=last_name, zipcode=zipcode
  File "/app/npsp.py", line 710, in get_or_create
    contact = cls.get(email=email)
  File "/app/npsp.py", line 760, in get
    response = sf.query(query)
  File "/app/npsp.py", line 119, in query
    self.check_response(r)
  File "/app/npsp.py", line 101, in check_response
    raise e
npsp.SalesforceException: Expected 200 but got 404
```

**Comment by AndrewGibson27** [comment]
WFM. If it's OK with you, I'd like to merge my customer-ID branch first (assuming it LGTY). I'll handle the inevitable merge conflicts here, checking with you if I have questions about how to resolve them.

**Comment by AndrewGibson27** [comment]
Also, I went to chicken.com because I was curious and was really sad someone hasn't done something more exciting with a prime domain like that.

File: forms.py [hunk 1]
```diff
@@ -1,48 +1,100 @@
...
+    openended_status = HiddenField(u"Openended Status")
+    campaign_id = HiddenField("Campaign ID")
+    referral_id = HiddenField("Referral ID")
+    description = HiddenField(u"Description")
+    pay_fees = BooleanField(u"Agree to pay fees")
+    pay_fees_value = HiddenField(u"Pay Fees Value")
+
+
+class BusinessMembershipForm(FlaskForm):
+    first_name = StringField(
```
[the remark this answers is not in the document]
Comment (x110dc) [reply forms.py]: Yes. I like it.

**Comment by x110dc** [comment]
Yes, merging that PR first is fine. Also given the changes to URL handling I think this endpoint should be something friendlier like `/business` maybe.

Sadly, chicken.com doesn't even support https.

**Pull request merged**